### DISCUSSION TITLE
Build issue with QNX

### DISCUSSION THREAD

**Discussion by GuruDlt** [libunwind/libunwind#636] [Q&A]
Hello Team,

I am trying to build unwind library with QNX with below configuration

./configure --host=aarch64-unknown-nto-qnx7.1.0 CFLAGS="-I/home/gta3kor/qnx710/target/qnx7/usr/include -I/home/gta3kor/qnx710/host/linux/x86_64/usr/lib/gcc/x86_64-pc-nto-qnx7.1.0/8.3.0/include -D__QNX__=1 -D__QNXNTO__=1 -D__aarch64__=1 -D__ORDER_LITTLE_ENDIAN__=1234 -D__FLOAT_WORD_ORDER__=__ORDER_LITTLE_ENDIAN__ -D__ORDER_PDP_ENDIAN__=3412 -D__LITTLEENDIAN__=1 -D__ORDER_BIG_ENDIAN__=4321 -D__BYTE_ORDER__=__ORDER_LITTLE_ENDIAN__ -D__EXT=1 -nostdinc" LDFLAGS="-L/home/gta3kor/qnx710/target/qnx7/usr/lib"

and I am facing below error:
mi/backtrace.c: In function ‘unw_backtrace’:
../include/libunwind-aarch64.h:245:94: **error: ‘struct unw_sigcontext’ has no member named ‘cpu’**
  245 | # define UNW_BASE register uint64_t unw_base __asm__ ("x0") = (uint64_t) unw_ctx->uc_mcontext.cpu.gpr;
      | 
[buid.log](https://github.com/libunwind/libunwind/files/12739223/buid.log)

can anyone please help me here on priority.

**Comment by GuruDlt** [comment]
this is the where code is falling when it is failing

```
#if defined(__FreeBSD__)
**_**_# define UNW_BASE register uint64_t unw_base __asm__ ("x0") = (uint64_t) unw_ctx->uc_mcontext.mc_gpregs.gp_x;_**_**
#elif defined(__QNX__)
**_# define UNW_BASE register uint64_t unw_base __asm__ ("x0") = (uint64_t) unw_ctx->uc_mcontext.cpu.gpr;_**
#else
**_**# define UNW_BASE register uint64_t unw_base __asm__ ("x0") = (uint64_t) unw_ctx->uc_mcontext.regs;**_**
#endif

#define unw_tdep_getcontext(uc) ({					\
  unw_tdep_context_t *unw_ctx = (uc);					\
  UNW_BASE \
  __asm__ __volatile__ (					        \
     "stp x0, x1, [%[base], #0]\n" \
     "stp x2, x3, [%[base], #16]\n" \
     "stp x4, x5, [%[base], #32]\n" \
     "stp x6, x7, [%[base], #48]\n" \
     "stp x8, x9, [%[base], #64]\n" \
     "stp x10, x11, [%[base], #80]\n" \
     "stp x12, x13, [%[base], #96]\n" \
     "stp x14, x15, [%[base], #112]\n" \
     "stp x16, x17, [%[base], #128]\n" \
     "stp x18, x19, [%[base], #144]\n" \
     "stp x20, x21, [%[base], #160]\n" \
     "stp x22, x23, [%[base], #176]\n" \
     "stp x24, x25, [%[base], #192]\n" \
     "stp x26, x27, [%[base], #208]\n" \
     "stp x28, x29, [%[base], #224]\n" \
     "mov x1, sp\n" \
     "stp x30, x1, [%[base], #240]\n" \
     "adr x1, ret%=\n" \
     "str x1, [%[base], #256]\n" \
     "mov %[base], #0\n" \
     "ret%=:\n" \
     : [base] "+r" (unw_base) : : "x1", "memory"); \
  (int)unw_base; })
```

**Comment by bregma** [comment]
Try `./configure --host=x86_64-pc-nto-qnx7.1.0 --build=x86_64-linux-gnu` instead of all that cruft. You may not be configuring correctly for cross building.

**Comment by GuruDlt** [comment]
Hi @bregma ,

Thanks for your reply and support.
I have tried as you said and it started throwing errors and stated from below error,

In file included from os-qnx.c:34:
In file included from /home/gta3kor/qnx710/target/qnx7/usr/include/sys/link.h:29,
                 from os-qnx.c:34:
/usr/include/x86_64-linux-gnu/sys/elf.h:22:3: **error** : #error This header is unsupported on x86-64.
   22 | # error This header is unsupported on x86-64.
      |   ^~~~~
In file included from os-qnx.c:34:
/home/gta3kor/qnx710/target/qnx7/usr/include/sys/link.h:41:8: **error** : redefinition of ‘struct link_map’

Please find the build log file respectively and kindly help me with the same.

[build_log_sep28_1.txt](https://github.com/libunwind/libunwind/files/12745908/build_log_sep28_1.txt)

**Comment by GuruDlt** [comment]
Hi @bregma ,

I request you to see once my build environment parameters which I am passing and kindly help me.

:~/clone/TEST_ACP/libunwind$ ./configure --host=aarch64-unknown-nto-qnx7.1.0 --build=x86_64-linux-gnu  CFLAGS="-I/home/gta3kor/qnx710/target/qnx7/usr/include -I/home/gta3kor/qnx710/host/linux/x86_64/usr/lib/gcc/x86_64-pc-nto-qnx7.1.0/8.3.0/include -D=**__aarch64__**=1 -D__QNX__=1 -D__QNXNTO__=1 -D__ORDER_LITTLE_ENDIAN__=1234 -D__FLOAT_WORD_ORDER__=__ORDER_LITTLE_ENDIAN__ -D__ORDER_PDP_ENDIAN__=3412 -D__LITTLEENDIAN__=1 -D__ORDER_BIG_ENDIAN__=4321 -D__BYTE_ORDER__=__ORDER_LITTLE_ENDIAN__ -D__EXT=1 -nostdinc" LDFLAGS="-L/home/gta3kor/qnx710/target/qnx7/usr/lib"

[build_log_sep28_2.txt](https://github.com/libunwind/libunwind/files/12748800/build_log_sep28_2.txt)

And with above mentioned configure command, I m facing below error,

mi/backtrace.c: In function ‘unw_backtrace’:
../include/libunwind-aarch64.h:246:37: **error** : **invalid register name for ‘unw_base’**
  246 | # define UNW_BASE register uint64_t unw_base __asm__ ("x0") = (uint64_t) unw_ctx->uc_mcontext.cpu.gpr;
      |                                     ^~~~~~~~
../include/libunwind-aarch64.h:253:3: note: in expansion of macro ‘UNW_BASE’
  253 |   UNW_BASE \

**Comment by bregma** [comment]
I see in your build log the line
```
checking for x86_64-pc-nto-qnx7.1.0-gcc... no
```
which looks to me like you have not set your QNX development environment up correctly because that binary should be in your **$PATH**. Since it's not, you're using your hosts's `gcc` binary and including host header files and libraries.

The right way to build for QNX would something like this.
```
  $ . /home/gta3kor/qnx710/qnxsdp-env.sh
  $ cd ~/clone/TEST_ACP/libunwind
  $ mkdir build-qnx
  $ cd build-qnx
  $ ../configure --host=x86_64-pc-nto-qnx7.1.0 --build=x86_64-linux-gnu
  $ make
```
You *must* set up your QNX (or other cross-compile) build environment correctly, and you *must* explicitly specify both `--build` and `--host` for configure to pick up the fact that's it's cross compiling. This is true of any and every cross-compile target. Otherwise you end up picking up the hosts's toolchain, headers, and libraries.

**Comment by GuruDlt** [comment]
Hello @bregma 

I have tried as suggested, see the below log and i could see its not successful.

Please find the attached build log and help me
[build_log_sep28_3.txt](https://github.com/libunwind/libunwind/files/12750336/build_log_sep28_3.txt)

**Comment by bregma** [comment]
Something is still wrong with your environment. When I run the commands the first part of the configure output is this.
```
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-nto-qnx7.1.0
checking target system type... x86_64-pc-nto-qnx7.1.0
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for x86_64-pc-nto-qnx7.1.0-strip... x86_64-pc-nto-qnx7.1.0-strip
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... no
checking for mawk... mawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether make supports nested variables... (cached) yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking for x86_64-pc-nto-qnx7.1.0-gcc... x86_64-pc-nto-qnx7.1.0-gcc
```
Your log shows this.
```
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-nto-qnx7.1.0
checking target system type... x86_64-pc-nto-qnx7.1.0
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for x86_64-pc-nto-qnx7.1.0-strip... no
checking for strip... strip
configure: WARNING: using cross tools not prefixed with host triplet
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... no
checking for mawk... mawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether make supports nested variables... (cached) yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking for x86_64-pc-nto-qnx7.1.0-gcc... no
```
The configure script looks in $PATH for the executable `x86_64-pc-nto-qnx7.1.0-gcc`. It's not finding it. That means either your $PATH is missing the QNX toolchain directory (and sourcing the `.qnxsdp-env.sh` file should set it correctly by prepending /home/gta3kor/sdp_710/host/linux/x86_64/usr/bin) or else the toolchain is missing from your QNX SDP 7.1 installation. Is it possible you're on Windows using WSL and have installed the Windows host tools instead of the Linux host tools?

You need to investigate why your environment is not correct, and you need to fix it in order to continue.

**Comment by GuruDlt** [comment]
Hi @bregma 

Sure I m doing my investigation but can you confirm one thing that,
with below parameter, were you tried to build for linux aarch64 or for QNX aarch64??

**Reply by bregma** [reply]
```
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-nto-qnx7.1.0
```
This says both you and I were configuring for an x86_64 GNU/Linux build host and an x86_64 QNX OS 7.1 target.

I know the QNX SDP does not support an aarch64 build host of any type (only x86_64 Windows, x86_64 GNU/Linux, and x86_64 Mac OS). I frequently test QNX aarch64 targets without problems -- just use `--host=aarch64-unknown-nto-qnx7.1.0`.

**Reply by GuruDlt** [reply]
Hi @bregma 

I will try as suggested with "--host=aarch64-unknown-nto-qnx7.1.0"
but in my case target is aarch64 QNX right, I am not trying to build for Qemu, I am trying to build for aarch64 HW.

**Comment by GuruDlt** [comment]
Hi @bregma 

I have tried as suggested  "--host=aarch64-unknown-nto-qnx7.1.0" as "**./configure --host=aarch64-unknown-nto-qnx7.1.0** ", but got below error

gta3kor@KOR-C-001Y3:~/clone/TEST_ACP/libunwind$ make
Making all in src
make[1]: Entering directory '/home/gta3kor/clone/TEST_ACP/libunwind/src'
make  all-am
make[2]: Entering directory '/home/gta3kor/clone/TEST_ACP/libunwind/src'
  CC       os-qnx.lo
os-qnx.c:33:10: fatal **error** : sys/link.h: No such file or directory
   33 | #include "sys/link.h"
      |          ^~~~~~~~~~~~

then I have ran the command, "**./configure --host=aarch64-unknown-nto-qnx7.1.0 --build=x86_64-linux-gnu  CFLAGS="-I/home/gta3kor/qnx710/target/qnx7/usr/include -I/home/gta3kor/qnx710/host/linux/x86_64/usr/lib/gcc/x86_64-pc-nto-qnx7.1.0/8.3.0/include"**"

then I got many error starting from

/home/gta3kor/qnx710/target/qnx7/usr/include/sys/compiler_gnu.h:60:3: **error** : #error Endian not defined
   60 | # error Endian not defined
   
   
  Please find respective build log the same and kindly help me.
[build_log_oct5_1.txt](https://github.com/libunwind/libunwind/files/12799922/build_log_oct5_1.txt)

**Reply by GuruDlt** [reply]
Hello @bregma 

I just have one question

-> what is the importance of  function **unw_tdep_getcontext** in the context of QNX aarch64
where and all above mentioned function has been used in case of use case

**Reply by bregma** [reply]
```
checking for aarch64-unknown-nto-qnx7.1.0-gcc... no
```
Changing your target architecture from x86_64 to aarch64le is not going to change the fact that your host tools are not being found.

The configure script looks in $PATH for the executable `aarch64-unknown-nto-qnx7.1.0-gcc`. It's not finding it. That means either your $PATH is missing the QNX toolchain directory (and sourcing the .qnxsdp-env.sh file should set it correctly by prepending /home/gta3kor/sdp_710/host/linux/x86_64/usr/bin) or else the toolchain is missing from your QNX SDP 7.1 installation. Is it possible you're on Windows using WSL and have installed the Windows host tools instead of the Linux host tools? Are you using an aarch64 development host (which is unsupported by QNX)?

You should get output similar to the following.
```
$ source ~/sdp_710/qnxsdp-env.sh
QNX_HOST=/home/swebb/sdp_710/host/linux/x86_64
QNX_TARGET=/home/swebb/sdp_710/target/qnx7
MAKEFLAGS=-I/home/swebb/sdp_710/target/qnx7/usr/include
$ find $QNX_HOST -name aarch64-unknown-nto-qnx7.1.0-gcc
/home/swebb/sdp_710/host/linux/x86_64/usr/bin/aarch64-unknown-nto-qnx7.1.0-gcc
```

If you do not, you need to investigate why your environment is not correct, and you need to fix it in order to continue. This has nothing to do with libunwind.

**Reply by GuruDlt** [reply]
Hello @bregma

Thanks for your support here.
I have one question

-> what is the importance of function **unw_tdep_getcontext** in the context of QNX aarch64
where and all above mentioned function has been used in case of use case

**Reply by GuruDlt** [reply]
Hello @bregma

Thanks for your support here.
I have one question

-> what is the importance of function **unw_tdep_getcontext** in the context of QNX aarch64
where and all above mentioned function has been used in case of use case

**Reply by bregma** [reply]
`unw_tdep_getcontext()` is the internal target-dependent implementation of `unw_getcontext()` specialized in this case for aarch64 and the target host OS ABI. Every OS kernel and CPU combination has its own unique binary interface to its internal processor context (CPU registers saved to memory between context switches). You generally call `unw_getcontext()` to get the current context from which to begin unwinding.

The API is [publicly documented](https://www.nongnu.org/libunwind/man/unw_getcontext(3).html).
